### PR TITLE
Initial implementation.

### PR DESCRIPTION
**Pull request by tchow8**
Initial implementation to:

Get data from Staging storage account between passed in fromDate and toDate.
Write data from JsonL format to CSV format.
Encrypt and zip the result and store in Export storage account.
Generate SAS link to download the blob.
Send email with link to relevant email targets.

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -59,11 +59,19 @@ stages:
       product: $(product)
       chartEnv: $(environment)
       preIntegrationTest: true
-      functionalTest: false
+      functionalTest: true
       azureVault: $(vault)
-      azureSecreKeys: 'mi-staging-storage-connection-string'
+      azureSecreKeys: 'mi-staging-storage-connection-string,mi-export-storage-connection-string,mail-username,mail-password,archive-password'
```
Comment (qzhou-hmcts) [inline azure-pipelines.yml]: do you need the connection strings?

Comment (tchow8) [reply azure-pipelines.yml]: Was testing with it. Will remove them once I verify the Managed Identity works now with the latest merge.

File: charts/mi-extraction-service/values.yaml [hunk 2]
```diff
@@ -1,3 +1,29 @@
...
+  aadIdentityName: mi-extraction-service
+  kind: Job
+  labels:
+    app.kubernetes.io/instance : mi-extraction-service-build-001
+    app.kubernetes.io/name: mi-extraction-service-build
   environment:
-    image: 'ssprivatesbox.azurecr.io/mi/mi-extraction-service:latest'
+    STORAGE_STAGING_NAME: mipersistent{{ .Values.global.environment }}
+    STORAGE_EXPORT_NAME: miexport{{ .Values.global.environment }}
+    MAIL_TARGETS: tony.chow@hmcts.net
```
Comment (qzhou-hmcts) [inline charts/mi-extraction-service/values.yaml]: probably you don't want to merge this

Comment (tchow8) [reply charts/mi-extraction-service/values.yaml]: Yep :+1:, I'll make it a secret instead. Easier when moving to GovUK Notify and I'll use the secret as an API Key isntead.

File: src/main/java/uk/gov/hmcts/reform/mi/miextractionservice/component/impl/CoreCaseDataExportBlobDataComponentImpl.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            if (blobContainerItem.getName().startsWith(CCD_DATA_CONTAINER_PREFIX)) {
+                for (BlobItem blobItem : sourceBlobServiceClient.getBlobContainerClient(blobContainerItem.getName()).listBlobs()) {
+                    if (blobNameIndexes.parallelStream().anyMatch(blobItem.getName()::contains)) {
+
+                        byte[] blobData = blobDownloadComponent
+                            .downloadBlob(sourceBlobServiceClient, blobContainerItem.getName(), blobItem.getName());
+
+                        List<String> stringData = readerUtil.readBytesAsStrings(blobData);
+
+                        outputData.addAll(convertDataStringsToOutput(stringData, fromDate, toDate));
```
Comment (qzhou-hmcts) [inline src/main/java/uk/gov/hmcts/reform/mi/miextractionservice/component/impl/CoreCaseDataExportBlobDataComponentImpl.java]: the method name is misleading, as it missing the filter part
probably is cleaner to this in two steps using streams
stringData.stream().filter(fromDate, toDate).map(toOutputCoreCaseData).collect()

File: src/main/java/uk/gov/hmcts/reform/mi/miextractionservice/component/impl/CoreCaseDataExportBlobDataComponentImpl.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            OffsetDateTime eventCreatedDate = OffsetDateTime.ofInstant(Instant.ofEpochMilli(coreCaseData.getCeCreatedDate()), ZoneOffset.UTC);
+            if (eventCreatedDate.isAfter(fromDate) && eventCreatedDate.isBefore(toDate)) {
+                outputCoreCaseDataList.add(coreCaseDataFormatterComponent.formatData(coreCaseData));
+            }
+        });
+
+        return outputCoreCaseDataList;
+    }
+
+    private List<String> getYearMonthIndexes(OffsetDateTime fromDate, OffsetDateTime toDate) {
```
Comment (qzhou-hmcts) [inline src/main/java/uk/gov/hmcts/reform/mi/miextractionservice/component/impl/CoreCaseDataExportBlobDataComponentImpl.java]: Just an alternative option :smiley:

```suggestion
    private List<String> getYearMonthIndexes(OffsetDateTime fromDate, OffsetDateTime toDate) {
          DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM");
        List<String> listOfIndexes = new ArrayList<>();
        OffsetDateTime current = fromDate.withDayOfMonth(1);
        while(current.isBefore(toDate)) {
            listOfIndexes.add(DATE_TIME_FORMATTER.format(current));
            current = current.plusMonths(1);
        }
        return listOfIndexes;
        }
```

**Comment by tchow8** [comment]
@qzhou-hmcts made changes. Please check.

**Comment by qzhou-hmcts** [comment]
LGTM

**Pull request merged**